google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1/RequestProfiler.java

```diff
@@ -0,0 +1,325 @@
+    // Time spent to fetch the table schema when user didn't provide it.
+    SCHEMA_FETCHING("schema_fetching"),
+    // Time spent within wait queue before it get picked up.
+    WAIT_QUEUE("wait_queue"),
+    // Time spent within backend to process the request.
+    NETWORK_RESPONSE_LATENCY("network_response_latency");
+    private final String operationName;
+
+    OperationName(String operationName) {
+      this.operationName = operationName;
+    }
+  }
+
+  private static final Logger log = Logger.getLogger(RequestProfiler.class.getName());
+
+  // Discard the requests if we are caching too many requests.
+  private static final int MAX_CACHED_REQUEST = 100000;
+
+  // Singleton for easier access.
+  public static final RequestProfiler REQUEST_PROFILER_SINGLETON = new RequestProfiler();
+
+  // Tunable static variable indicate how many top longest latency requests we should consider.
+  private static int TOP_K = 10;
+
+  // Tunable static variable indicate how often the report should be generated.
+  private static Duration FLUSH_PERIOD = Duration.ofMinutes(1);
+
+  // From request uuid to the profiler of individual request. This will be cleaned up periodically.
+  private final Map<String, IndividualRequestProfiler> idToIndividualOperation =
+      new ConcurrentHashMap<>();
+
+  private Thread flushThread;
+
+  // Count the total number of dropped operations.
+  AtomicLong droppedOperationCount = new AtomicLong(0);
+
+  // Mark an operation for a given request id to be start.
+  void startOperation(OperationName operationName, String requestUniqueId) {
+    if (!idToIndividualOperation.containsKey(requestUniqueId)) {
+      if (idToIndividualOperation.size() > MAX_CACHED_REQUEST) {
+        log.warning(
+            String.format(
+                "startOperation is triggered for request_id: %s that's hasn't "
+                    + "seen before, this is possible when "
+                    + "we are recording too much ongoing requests. So far we has dropped %s operations.",
+                requestUniqueId, droppedOperationCount));
+        droppedOperationCount.incrementAndGet();
+        return;
+      }
+      idToIndividualOperation.put(requestUniqueId, new IndividualRequestProfiler(requestUniqueId));
+    }
+    idToIndividualOperation.get(requestUniqueId).startOperation(operationName);
+  }
+
+  // Mark an operation for a given request id to be end.
+  void endOperation(OperationName operationName, String requestUniqueId) {
+    if (!idToIndividualOperation.containsKey(requestUniqueId)) {
+      log.warning(
+          String.format(
+              "endOperation is triggered for request_id: %s that's hasn't "
+                  + "seen before, this is possible when "
+                  + "we are recording too much ongoing requests. So far we has dropped %s operations.",
+              requestUniqueId, droppedOperationCount));
+      droppedOperationCount.incrementAndGet();
```

Comment:
Should this be incremented here as well? What if the calling code always calls startOperation() and endOperation()?

Reply:
Done